### ISSUE TITLE
Share code between hooks with common args

### ISSUE THREAD

**Issue by Andreagit97** [bpftrace/bpftrace#3590]
### Is your feature request related to a problem? Please describe.

As reported here https://github.com/bpftrace/bpftrace/discussions/3589#discussioncomment-11290641

I'm hooking into `kretfunc:tcp_sendmsg` and `kretfunc:tcp_recvmsg` and my 2 programs share almost the same code (this is just an example):

```
#!/usr/bin/bpftrace

kretfunc:tcp_sendmsg
{
    $inet = (struct inet_sock *)args->sk;
    $lport = bswap($inet->inet_sport);
    $dport = bswap(args->sk->__sk_common.skc_dport);

    printf("%d<-%d\n", $dport, $lport);
}

kretfunc:tcp_recvmsg
{
    $inet = (struct inet_sock *)args->sk;
    $lport = bswap($inet->inet_sport);
    $dport = bswap(args->sk->__sk_common.skc_dport);

    printf("%d->%d\n", $dport, $lport);
}
```

I would like to have a way to share the code between these 2 functions without duplicating it.
Today if i try to do something like this

```
kretfunc:tcp_sendmsg,
kretfunc:tcp_recvmsg
{
    $inet = (struct inet_sock *)args->sk;
    $lport = bswap($inet->inet_sport);
    $dport = bswap(args->sk->__sk_common.skc_dport);

    printf("%d<-%d\n", $dport, $lport);
}
```

bpftrace returns an error `ERROR: Probe has attach points with mixed arguments`. 
This is correct, but in the end, I'm just using an argument `args->sk` that both attach points have, so there shouldn't be issues

### Describe the solution you'd like

I would like to do something like this:

```
kretfunc:tcp_sendmsg,
kretfunc:tcp_recvmsg
{
    $inet = (struct inet_sock *)args->sk;
    $lport = bswap($inet->inet_sport);
    $dport = bswap(args->sk->__sk_common.skc_dport);

    if (probe == "tracepoint:syscalls:sys_enter_open") {
        printf("%d<-%d\n", $dport, $lport);
    } else if (probe == "tracepoint:syscalls:sys_exit_open") {
        printf("%d->%d\n", $dport, $lport);
    }
}
```

### Describe alternative solutions or features you've considered

https://github.com/bpftrace/bpftrace/issues/2516 seems a possible solution to this issue

**Comment by danobi** [comment]
Hi @Andreagit97 , like you found, user defined functions (UDF) is our long term solution for this.

That being said, I think your script is valid and _should_ be able to load. I'm not 100% sure if the kernel will block it, so a little more research might be required. If kernel is ok with this, we'd likely take a PR to relax restriction from bpftrace side. It's probably conservative to just keep things simple.

**Comment by viktormalik** [comment]
Kernel will be ok b/c the BPF programs (probes) are duplicated anyways. What's tricky is that we need to check that `args.sk` refers to the same argument position in both probes or otherwise emit a different codegen for the two attach points. But for fexit probes we do that anyways (at the moment) so just lifting off the restriction might actually work here. Let me have a look.

**Comment by viktormalik** [comment]
So, lifting off the restriction is not as straightforward as it may seem. At the moment, we always assume that each probe has a unique set of arguments, if it doesn't then we fail with the above error. This allows us to store arguments in a special struct type named after the probe.

In order to fix this issue, several things must be done:
1. Probe arguments must be stored per-attachpoint (instead of per-probe).
2. For probes with attach points with mixed arguments, we have multiple options:
    1. Allow arguments with the same name to have a different type across attach points. This will require to run the entire `FieldAnalyser` on each attach point (instead of just once per probe as we do now).
    2. Require arguments with the same name to be of the same type (which would suffice for this case). This will require to check this property while parsing arguments for a probe (and its attach points).

Whichever way we choose, all of these points require some fundamental changes to the way `FieldAnalyser` works. Normally I wouldn't have problem to do that but as @ajor noticed in #3422, `FieldAnalyser` is broken and needs complete rewriting, ideally merging with type propagation in semantic analyser (#2979). So, I'm a bit hesitant to go down this way as whatever we implement will be thrown away and rewritten later. Especially when there is a workaround for this issue.

**Comment by jordalgo** [comment]
@Andreagit97 Would the new macro syntax be able to help with this? Not prefect but should prevent some code duplication: https://github.com/bpftrace/bpftrace/pull/4037/files

**Comment by Andreagit97** [comment]
Yes, I think this should help. Thank you for sharing! We can close this issue on my side

**Comment by viktormalik** [comment]
We're also discussing moving probe expansion earlier in the bpfscript compilation, preferably before the field and semantic analyser passes. With that, we could probably detect a situation like this and perform full probe expansion (basically clone the probe for every attach point).
